File: leetcode-04/solutions/python3/0907.py
class Solution:
    def sumSubarrayMins(self, A: List[int]) -> int:
        n = len(A)
        kMod = int(1e9 + 7)
        
        ans = 0
        prev = [0] * n
        next = [0] * n
        stack1 = []
        stack2 = []
        
        for i in range(n):
            prev[i] = i + 1
            next[i] = n - i
        
        for i in range(n):
            while stack1 and A[stack1[-1]] > A[i]:
                stack1.pop()
            prev[i] = i - stack1[-1] if stack1 else i + 1
            stack1.append(i)
            
            while stack2 and A[stack2[-1]] > A[i]:
                next[stack2[-1]] = i - stack2[-1]
                stack2.pop()
            stack2.append(i)
        
        for i in range(n):
            ans = (ans + A[i] * prev[i] * next[i]) % kMod
        
        return ans
        

File: leetcode-04/solutions/cpp/0907.cpp
class Solution {
 public:
  int sumSubarrayMins(vector<int>& A) {
    const int n = A.size();
    const int kMod = 1e9 + 7;

    int ans = 0;
    vector<int> prev(n);
    vector<int> next(n);
    stack<int> stack1;
    stack<int> stack2;

    for (int i = 0; i < n; i++) {
      prev[i] = i + 1;
      next[i] = n - i;
    }

    for (int i = 0; i < n; i++) {
      while (!stack1.empty() && A[stack1.top()] > A[i]) stack1.pop();
      prev[i] = stack1.empty() ? i + 1 : i - stack1.top();
      stack1.push(i);

      while (!stack2.empty() && A[stack2.top()] > A[i]) {
        next[stack2.top()] = i - stack2.top();
        stack2.pop();
      }
      stack2.push(i);
    }

    for (int i = 0; i < n; i++) ans = (ans + A[i] * prev[i] * next[i]) % kMod;

    return ans;
  }
};